### DISCUSSION TITLE
Nullable generic not consistant with primitive and object type

### DISCUSSION THREAD

**Discussion by oli-unreal** [dotnet/csharplang#6139] [General] [answered]
C# version: 10

Given this code;
```
public static void AddNoNullTo<T>(this T? value, List<T> list)
{
    if (value == null) return;
    list.Add(value);
}
```

It works fine with this code:
```
public void ShouldNotAddNullValue()
{
    var list = new List<string> { "obj" };
    ((string?)null).AddNoNullTo(list);
    list.ShouldBe(new List<string>{ "obj" });
}
```

But it doesn't compile when using a primitive type:
```
public void ShouldNotAddNullValue()
{
    var list = new List<int> { 1 };
    ((int?)null).AddNoNullTo(list);
    list.ShouldBe(new List<int>{ 1 });
}
```

> Error CS0411 : The type arguments for method 'ListExtension.AddNoNullTo<T>(T?, List<T>)' cannot be inferred from the usage. Try specifying the type arguments explicitly.

I was not expecting compilation issue depending of the type of the generic

**Comment by HaloFour** [comment, marked as answer]
Nullability works differently between reference types and value types (including primitives) and they are not compatible with one another.  `string?` is really just `string`, but `int?` is actually `Nullable<int>`.

**Reply by svick** [reply]
This also means that you can add an overload of `AddNoNullTo` specifically for nullable value types:

```c#
public static void AddNoNullTo<T>(this T? value, List<T> list)
    where T : struct
{
    if (value == null) return;
    list.Add(value.Value);
}
```
